xtask/src/ai.rs

```diff
@@ -0,0 +1,109 @@
+use test_context::AsyncTestContext;
+use trustify_module_fundamental::ai::model::ChatState;
+use trustify_module_fundamental::ai::service::AiService;
+use trustify_test_context::TrustifyContext;
+
+#[derive(Debug, Parser, Default)]
+pub struct Ai {}
+
+impl Ai {
+    pub async fn run(self) -> anyhow::Result<()> {
+        run().await
+    }
+}
+
+async fn run() -> anyhow::Result<()> {
+    // Set the following environment variables to against a different OpenAI API endpoint
+    // OPENAI_API_BASE, OPENAI_API_KEY, OPENAI_MODEL
+
+    // export EXTERNAL_TEST_DB to use an external database, otherwise
+    // a temporary database will be created for testing.
+    let ctx = TrustifyContext::setup().await;
+
+    // if we are not using an external database, ingest some documents so that
+    // we use the data in them to respond to questions.
+    if env::var("EXTERNAL_TEST_DB").is_err() {
+        println!("{}", LightGray.paint("ingesting documents..."));
+        _ = ctx
+            .ingest_document("quarkus-bom-2.13.8.Final-redhat-00004.json")
+            .await?;
+        _ = ctx.ingest_document("ubi9-9.2-755.1697625012.json").await?;
+        _ = ctx
+            .ingest_document("zookeeper-3.9.2-cyclonedx.json")
+            .await?;
+        _ = ctx.ingest_document("cve/CVE-2021-32714.json").await?;
+        _ = ctx.ingest_document("cve/CVE-2024-26308.json").await?;
+        _ = ctx.ingest_document("cve/CVE-2024-29025.json").await?;
+        _ = ctx.ingest_document("csaf/cve-2023-33201.json").await?;
+        println!("{}", Green.paint("DONE!"));
+    }
+
+    let service = AiService::new(ctx.db.clone());
+
+    // let question = "What the latest version of quarkus-bom?";
+    // println!("question: {}", Blue.paint(question));
+    // ask(&executor, question).await?;
+
+    let history = Box::new(FileBackedHistory::with_file(5, ".ai_history.txt".into())?);
+
+    let mut line_editor = Reedline::create().with_history(history);
+
+    let prompt = DefaultPrompt {
+        left_prompt: DefaultPromptSegment::Basic("Trustify Assistant".to_string()),
+        right_prompt: DefaultPromptSegment::Basic(">>".to_string()),
+    };
+
+    let mut chat_state = ChatState::new();
+
+    loop {
+        match line_editor.read_line(&prompt) {
+            Ok(Signal::Success(buffer)) => {
+                match buffer.trim().to_lowercase().as_str() {
+                    "quit" | "exit" => {
+                        println!("\nBye!");
+                        return Ok(());
+                    }
+                    "clear" => {
+                        chat_state = ChatState::new();
+                        println!("\nChat history cleared...");
+                        continue;
+                    }
+                    _ => {}
+                }
+
+                chat_state.add_human_message(buffer.clone());
+                let pos = chat_state.messages.len();
+
+                let new_state = service.completions(&chat_state, ()).await?;
```

Comment:
What's the xtask for? Is it just another way of accessing the same service accessible through `/api/v1/ai/completions`?

Reply:
Right.  a CLI version of it.